### PR TITLE
Lambda concurrency check

### PR DESCRIPTION
**Pull request by paranshu-singhal**
I confirm these files are made available under CC0 1.0 Universal (https://creativecommons.org/publicdomain/zero/1.0/legalcode)

*Issue #, if available:*

*Description of changes:*

### CODE REVIEW

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 1]
```diff
@@ -10,64 +10,65 @@
 # the specific language governing permissions and limitations under the License.
 
 '''
+Rule Name:
+  LAMBDA_CONCURRENCY_CHECK
 Description:
   Checks whether the AWS Lambda function is configured for function-level concurrent execution limit. The rule is NON_COMPLIANT if the Lambda function is not configured for function-level concurrent execution limit.
-
```
Comment (santosh07bec) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Please maintain the Gherkin Formatting.
No need to remove these blank lines.
Please format the Gherkin as in below rule Gherkin:
https://github.com/awslabs/aws-config-rules/blob/master/python/ELASTICACHE_REDIS_CLUSTER_AUTO_BACKUP_CHECK/ELASTICACHE_REDIS_CLUSTER_AUTO_BACKUP_CHECK.py

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 2]
```diff
@@ -10,64 +10,65 @@
 # the specific language governing permissions and limitations under the License.
 
 '''
+Rule Name:
+  LAMBDA_CONCURRENCY_CHECK
 Description:
```
Comment (santosh07bec) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Add a empty line here.

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 3]
```diff
@@ -10,64 +10,65 @@
...
 
 '''
+Rule Name:
+  LAMBDA_CONCURRENCY_CHECK
 Description:
   Checks whether the AWS Lambda function is configured for function-level concurrent execution limit. The rule is NON_COMPLIANT if the Lambda function is not configured for function-level concurrent execution limit.
-
 Trigger:
-  Periodic
-
```
Comment (santosh07bec) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Same here

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 4]
```diff
@@ -10,64 +10,65 @@
...
+  Configuration change on AWS::Lambda::Function
 Reports on:
   AWS::Lambda::Function
-
```
Comment (santosh07bec) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Here as well.

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 5]
```diff
@@ -10,64 +10,65 @@
...
-  concurrencyLimitLow
-  (Optional) To check if the function concurrency is less than or equal to what has been specified in the concurrencyLimitLow
-
-  concurrencyLimitHigh
-  (Optional) To check if the function concurrency is higher than or equal to what has been specified in the concurrencyLimitHigh
-
-
+  ConcurrencyLimitLow
+  (Optional) To check if the function concurrency is higher than or equal to what has been specified in the ConcurrencyLimitLow
+  ConcurrencyLimitHigh
```
Comment (santosh07bec) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Same

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 6]
```diff
@@ -10,64 +10,65 @@
...
+  (Optional) To check if the function concurrency is less than or equal to what has been specified in the ConcurrencyLimitHigh
 Scenarios:
```
Comment (santosh07bec) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: same

File: python/LAMBDA_CONCURRENCY_CHECK/parameters.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{
+  "Version": "1.0",
+  "Parameters": {
+    "RuleName": "LAMBDA_CONCURRENCY_CHECK",
+    "SourceRuntime": "python3.6",
+    "CodeKey": "LAMBDA_CONCURRENCY_CHECK.zip",
+    "InputParameters": "{}",
+    "OptionalParameters": "{}",
```
Comment (jongogogo) [inline python/LAMBDA_CONCURRENCY_CHECK/parameters.json]: Add the parameters as keys with an empty value.

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+
+    function_name = configuration_item['configuration']['functionName']
+    if 'Concurrency' not in configuration_item['supplementaryConfiguration']:
+        return build_evaluation(function_name, 'NON_COMPLIANT', event,
```
[the remark this answers is not in the document]
Comment (santosh07bec) [reply python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: DOne!

File: python/LAMBDA_CONCURRENCY_CHECK/parameters.json (hunk 7 above)
Comment (santosh07bec) [reply python/LAMBDA_CONCURRENCY_CHECK/parameters.json]: Done!

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+
+    function_name = configuration_item['configuration']['functionName']
+    if 'Concurrency' not in configuration_item['supplementaryConfiguration']:
+        return build_evaluation_from_config_item(
+            configuration_item,
+            'NON_COMPLIANT',
+            annotation='Concurrency not set for the lambda function: {}'.format(function_name)
```
Comment (jongogogo) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: The function_name is already displayed in the resource id. No need to mention the functino name in the annotations

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+        return build_evaluation_from_config_item(configuration_item, 'COMPLIANT')
+
+    concurrency = configuration_item['supplementaryConfiguration']['Concurrency']['reservedConcurrentExecutions']
+
+    if 'ConcurrencyLimitLow' in valid_rule_parameters and 'ConcurrencyLimitHigh' not in valid_rule_parameters:
+        if concurrency < valid_rule_parameters['ConcurrencyLimitLow']:
+            return build_evaluation_from_config_item(
+                configuration_item,
+                'NON_COMPLIANT',
+                annotation='Concurrency of AWS Lambda function {} is lower then {}.'.format(
```
Comment (jongogogo) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: idem on function_name

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+                    )
+                )
+        return build_evaluation_from_config_item(configuration_item, 'COMPLIANT')
+
+    if 'ConcurrencyLimitHigh' in valid_rule_parameters and 'ConcurrencyLimitLow' not in valid_rule_parameters:
+        if concurrency > valid_rule_parameters['ConcurrencyLimitHigh']:
+            return build_evaluation_from_config_item(
+                configuration_item,
+                'NON_COMPLIANT',
+                annotation='Concurrency of AWS Lambda function {} is higher then {}.'.format(
```
Comment (jongogogo) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: idem on function_name

Comment (jongogogo) [reply python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: `higher than` not `then`

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+    if 'ConcurrencyLimitLow' in rule_parameters:
+        if int(rule_parameters['ConcurrencyLimitLow']) <= 0:
+            raise ValueError('ConcurrencyLimitLow must be a positive integer greater than 0.')
+        rule_parameters['ConcurrencyLimitLow'] = int(rule_parameters['ConcurrencyLimitLow'])
+    if 'ConcurrencyLimitHigh' in rule_parameters:
+        if int(rule_parameters['ConcurrencyLimitHigh']) <= 0:
+            raise ValueError('ConcurrencyLimitHigh must be a positive integer greater than 0.')
+        rule_parameters['ConcurrencyLimitHigh'] = int(rule_parameters['ConcurrencyLimitHigh'])
+    if all(key in rule_parameters for key in ['ConcurrencyLimitLow', 'ConcurrencyLimitHigh']) and \
+        rule_parameters['ConcurrencyLimitLow'] >= rule_parameters['ConcurrencyLimitHigh']:
```
Comment (jongogogo) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Shoud be `>` not `>=`. If it is equal it mandates a specific value.

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK_test.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+                         "configurationItemCaptureTime": "2019-04-18T08:49:09.878Z",\
+                         "supplementaryConfiguration": {"Concurrency":\
+                         {"reservedConcurrentExecutions": ' + str(concurrency) + '}}, "configurationItemStatus": "OK",\
+                         "configuration": {"functionName": "ABC"}}}'
+    return '{"notificationCreationTime": "asd","messageType": "ConfigurationItemChangeNotification",\
+                         "configurationItem": {"resourceType": "AWS::Lambda::Function",\
+                         "resourceId": "ABC",\
+                         "configurationItemCaptureTime": "2019-04-18T08:49:09.878Z",\
+                         "supplementaryConfiguration": {}, "configurationItemStatus": "OK",\
+                         "configuration": {"functionName": "ABC"}}}'
```
Comment (jongogogo) [inline python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK_test.py]: Clean up as not needed.

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py (hunk 9 above)
Comment (santosh07bec) [reply python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Done!

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py (hunk 10 above)
Comment (santosh07bec) [reply python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Done!

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py (hunk 11 above)
Comment (santosh07bec) [reply python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Done!

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py (hunk 12 above)
Comment (santosh07bec) [reply python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK.py]: Done!

File: python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK_test.py (hunk 13 above)
Comment (santosh07bec) [reply python/LAMBDA_CONCURRENCY_CHECK/LAMBDA_CONCURRENCY_CHECK_test.py]: Done

**Pull request merged**